Fixed #24747 -- Allowed transforms in QuerySet.order_by()

MatthewWilkes opened the pull request:
This is a PR to fix #24747, and other related tickets. It allows order_by to include results of arbitrary transforms.

This works by making setup_joins understand transforms and return an additional value that converts a fiel object to something the SQL compiler can understand that takes the transforms into account.

Practically, this allows order_by and distinct on JSONField, HStoreField and ArrayField.

MatthewWilkes commented on django/db/models/sql/compiler.py:
```diff
@@ -651,7 +651,9 @@ def find_ordering_name(self, name, opts, alias=None, default_order='ASC',
                                                        order, already_seen))
             return results
         targets, alias, _ = self.query.trim_joins(targets, joins, path)
-        return [(OrderBy(t.get_col(alias), descending=descending), False) for t in targets]
```
I would appreciate some guidance here, @jarshwah @charettes (and anyone else..)

It's my understanding that all the members of `targets` and `final_field` are guaranteed to have the same value, so I don't understand why we're looping over targets and generating `OrderBy` objects. It seems to me that if there was a reason we needed to generate multiple `ORDER BY` clauses then the `final_field` should be part of this too. It seems that just using `final_field` and not `targets` produces correct results, but I fear there is a subtlety of joins that I'm missing.

MatthewWilkes replied:
I've played some more here and I think the code ends up too messy this way. I'm now exploring returning a function as part of setup_joins that returns an object the compiler can handle. For fields this is equivalent to get_col. This should get around the problems around joins I was worried about. I'll push a change once I'm home from pycon.

MatthewWilkes replied:
To get this to work, I needed to allow `get_distinct` and `distinct_sql` to return a result, params tuple, rather than just a result, but I think this is a better approach. It's allowed me to get `DISTINCT ON` working with JSON keys. To get that, though, I had to add an `as_text` transform to get access to `#>>`, as JSON keys aren't orderable, unlike JSONB keys.

I'm still playing, and the branch is dirty, but this has got rid of my concerns above, so I think I'm on the right track.

pope1ni commented on django/db/models/sql/query.py:
```diff
@@ -1401,9 +1402,42 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
+            try:
+                path, final_field, targets, rest = self.names_to_path(
+                    names[:pivot], opts, allow_many, fail_on_missing=True)
+            except (FieldError) as exc:
```
Drop the unneeded parentheses.

MatthewWilkes replied:
Thank you, but as I said in the description I'm lokking for help with the mewt of the problem right now, not style advice. This does currently contain WIP code that is not suitable for review.

pope1ni commented on django/db/models/sql/query.py:
```diff
@@ -1401,9 +1402,42 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
+                    final_field = self.try_transform(final_field, name)
+                else:
+                    final_field = self.try_transform(final_field.get_col(alias), name)
+            except (FieldError):
```
Drop the unneeded parentheses.

pope1ni commented on django/db/models/sql/query.py:
```diff
@@ -1401,9 +1402,42 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
+                if hasattr(final_field, 'resolve_expression'):
+                    final_field = self.try_transform(final_field, name)
+                else:
+                    final_field = self.try_transform(final_field.get_col(alias), name)
```
```python
if not hasattr(final_field, 'resolve_expression'):
    final_field = final_field.get_col(alias)
try:
    final_field = self.try_transform(final_field, name)
    ...
```

MatthewWilkes replied:
Your suggestion here is actually wrong, it needs to be in a temporary variable, not final_field, otherwise if an exception is raised final_field is left as a Col rather than a Field.

MatthewWilkes commented:
The next things I'll be doing is adding the tests for HStore, ArrayField and JSONB lookups in postgres, mirroring the JSON ones.  If anyone has advice on other things to be tested, or if they think my approach has flaws I'd very much like to hear. As it stands, I'm happy with the approach of adding a transform function as a return from setup_joins.

MatthewWilkes commented:
buildbot, test on oracle

felixxm commented:
buildbot, test on oracle

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

jarshwah commented on django/db/models/sql/query.py:
```diff
@@ -1401,10 +1403,50 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
+        for pivot in range(len(names), 0, -1):
+            try:
+                path, final_field, targets, rest = self.names_to_path(
+                    names[:pivot], opts, allow_many, fail_on_missing=True)
```
I think we prefer the closing paren on a newline

jarshwah commented on django/db/models/sql/query.py:
```diff
@@ -1401,10 +1403,50 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
+                    # message, but we're going to reraise an earlier exception
+                    # so we can swallow this.
+                    if isinstance(final_field, Field) and last_field_exception:
+                        raise last_field_exception
```
I'm leaving a comment for me to check later. Does it make sense to raise current exception from last_field_exception? I think that'll show the name not being a transform caused by the name not being a field.

MatthewWilkes replied:
Hm, maybe. I think that perhaps we should raise it with a different context depending on if it's the first transform in the chain or not. I suspect that users are more likely to typo a field name or a join than they are a transform, so we'd want to push them to the (very helpful) `django.core.exceptions.FieldError: Cannot resolve keyword 'notafield' into field. Choices are: article, email, first_name, id, last_name` type errors we generate.

That said, if someone is doing ``datetimefield__date__day`` we don't really want to say that `date` isn't a field if they typo `day`.

jarshwah commented on django/db/models/sql/query.py:
```diff
@@ -1401,10 +1403,50 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
+                transforms = names[pivot:]
+                break
+        for name in transforms:
+            def transform(field, alias, name, previous):
```
`def transform(field, alias, *, name, previous)` ?

Makes the partial application further down more obvious that name and previous are supplied by keyword args. Not certain this is better, but I had to double check the partial applicability worked correctly.

MatthewWilkes replied:
Agreed.

jarshwah commented on django/db/models/sql/query.py:
```diff
@@ -1418,7 +1460,7 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
             alias = self.join(connection, reuse=reuse)
             joins.append(alias)
-        return final_field, targets, opts, joins, path
+        return final_field, targets, opts, joins, path, final_transformer
```
This is getting out of hand, especially considering callers ignoring so many of the return values. Does it make sense to create a type to model these values? At a minimum I think we should consider namedtuple.

MatthewWilkes replied:
This was done as #8840

jarshwah commented on tests/custom_lookups/tests.py:
```diff
@@ -63,6 +63,15 @@ def as_sql(self, compiler, connection):
 
 
+class LastDigitBilateralTransform(models.Transform):
+    bilateral = True
```
Does this make a difference to the tests? I'm not a fan of the bilateral attribute - it was introduced before expressions were a thing. Not sure we want to propagate their use.

MatthewWilkes replied:
Removed `bilateral` from this transform.

MatthewWilkes commented:
> .values('datefield__month') supported?

It is, I'll add an explicit test. `.values` is demonstrated in the postgres tests too, as:

```python
   def test_deep_values(self):
       query = JSONModel.objects.values_list('field__k__l')
       self.assertSequenceEqual(
           query,
           [(None, ), (None, ), (None, ), (None, ), (None, ), (None, ),
            (None, ), (None, ), ('m', ), (None, ), (None, ), (None, )]
       )
```

but I agree it should be in the platform agnostic suite.

> Any tests for multiple transformations? datefield__month__lastdigit or datefield__date__day?

Good idea.

> Any distinct tests?

Yeah, there are some in the postgres suite. I didn't do any in the backend agnostic as it's only really helpful to do it on distinct on, I think?

```python
   def test_deep_distinct(self):
       query = JSONModel.objects.distinct('field__k__l__as_text').values_list('field__k__l')
       self.assertSequenceEqual(
           query,
           [('m', ), (None, )]
       )
```

I added the `as_text` transform that works the same way as the JSON `KeyTransform` in that it merges multiple transforms together to give the right output. Using as_text at the end causes the `->>` operator to be used instead of `->`, which then compares the text representation rather than the data. This is needed as you don't have a general guarantee of having equality operators defined for the JSON field.

I should document that, too.

MatthewWilkes commented on django/db/models/sql/query.py:
```diff
@@ -1418,7 +1460,7 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
             alias = self.join(connection, reuse=reuse)
             joins.append(alias)
-        return final_field, targets, opts, joins, path
+        return final_field, targets, opts, joins, path, final_transformer
```
No argument here, I got a bit of underscore blindness at times. Think that should be its own PR, so we don't have too much refactoring in with functionality changes?

timgraham replied:
Yes please

MatthewWilkes commented on tests/custom_lookups/tests.py:
```diff
@@ -63,6 +63,15 @@ def as_sql(self, compiler, connection):
 
 
+class LastDigitBilateralTransform(models.Transform):
+    bilateral = True
```
It doesn't, I only did it to match with `mul3` and `div3`, as that seemed the closest match in the test suite for what I was trying to demonstrate.

timgraham commented:
It would be nice to update the docs/commit message/ticket, etc. to reflect the fact that `JSONField` `values_list` functionality. If the changes aren't coupled, even a separate commit for that may make review easier.

sir-sigurd commented on django/contrib/postgres/fields/jsonb.py:
```diff
@@ -118,6 +118,24 @@ class KeyTextTransform(KeyTransform):
     _output_field = TextField()
 
 
+class CoerceToText(Transform):
```
I think that this should be added in the separate pull request.

MatthewWilkes replied:
I've removed this functionality, it seems the tests pass with it gone, so either something else changed in the interim or I didn't realise I'd changed something elsewhere.

charettes commented on django/contrib/postgres/fields/jsonb.py:
```diff
@@ -118,6 +118,24 @@ class KeyTextTransform(KeyTransform):
     _output_field = TextField()
 
 
+class CoerceToText(Transform):
```
Also, I think this can be achieved by defining `KeyTextTransform.lookup_name` and registering it as a lookup. The reason why this wasn't done in the first place is because of the possible `lookup_name` (e.g. `'as_text'`) and jsonb key collision. See [#27257](https://code.djangoproject.com/ticket/26511) for more details.

MatthewWilkes replied:
Hm, yeah, these are good pointers. The problem I had that initially made me add it was to make sure the value returned in postgres was fully orderable, as just using the normal `KeyTransform` was causing errors with comparitor operations being undefined. That doesn't really make sense to me, as I thought that was one of the things JSONB was supposed to solve.

Does it make sense to have lookups that convert various things that could be in JSONFields to other Django primitive fields, so we can chain the lookups?

Splitting the PR out makes sense, I've been splitting the `setup_joins` stuff out already, so can do this one too.

MatthewWilkes commented:
Just want to follow up and apologise for not moving this on for weeks now. I've not forgotten, I'm going to try and find some time this week or next to get all the comments addressed.

MatthewWilkes commented on django/db/models/sql/compiler.py:
```diff
@@ -605,21 +610,22 @@ def get_distinct(self):
                 else:
-                    result.append("%s.%s" % (qn(alias), qn2(target.column)))
-        return result
+                    r, p = transform_function(target, alias).as_sql(self, self.connection)
```
I'm not sure how this would change to do that, does `as_sql` not take the backend into account? I can't find reference to `as_backend()` in the codebase I'm afraid.

jarshwah replied:
`.as_sql()` is the default implementation. `as_{backend.vendor}` is the backend specific implementation *if it exists*.

`compiler.compile` checks if a backend specific implementation exists first and uses that, falling back to `as_sql` if one does not exist.

So, `self.compile(transform_function(..))` I think would be correct.

MatthewWilkes replied:
Ah, thanks! I'll make this change.

MatthewWilkes commented on django/db/backends/base/operations.py:
```diff
@@ -151,7 +151,7 @@ def distinct_sql(self, fields):
             raise NotImplementedError('DISTINCT ON fields is not supported by this database backend')
         else:
-            return 'DISTINCT'
+            return ['DISTINCT'], []
```
According to the MSSQL backend docs, only one Django version is supported at a time. I think that should make it trivial, but your suggestion works well too. I think this is a safe change.

MatthewWilkes commented:
> I'd like to see more tests outside contrib/postgres that verify order/values/distinct/f usage. It looks like most of the cases above are actually tested within the json/hstore tests, but that's no good for all other backends.

`distinct on` is a harder one to test, as many backends don't support it. We have `test_transform_order_by` that orders by `age__lastdigit` and the `FLookupTests` class that do values and aggregate. I can write more, but I think these cover the set of functionality so I don't know what extra things to add that would be testing things that aren't currently covered.

> If .objects.values('datefield__month').annotate(sales=Count('pk')) now works (without the additional annotate preceding the values clause) I'd love to see that as a test, and to write up some usage docs showing the ability to write those kinds of queries.

I don't follow this one, I'm afraid?

MatthewWilkes commented:
I've added a little more in the way of documentation and gone through looking for inconsistencies, but I still have the question from the above comment in October.

jarshwah commented:
> I don't follow this one, I'm afraid?

Now that `field__transform` works in `order_by()`, I'm wondering if it works in `.values()` as well by happenstance.

If you want to group by a transform right now, you have to do this:

```
(
    Model
    .objects
    .annotate(transformed=Transform('field'))
    .values('transformed')
    .annotate(thecount=Count('id'))
)
```

What I want to be able to do is this:

```
(
    Model
    .objects
    .values('field__transform')
    .annotate(thecount=Count('id'))
)
```

I was hoping that the use of `transform_function` from JoinInfo will *just work* for `values()` as well as `order_by()`.

MatthewWilkes commented:
@jarshwah Regarding `.values()`, that is indedd the case:

Documentation patch is at:
https://github.com/django/django/pull/8528/files#diff-5e313dd49d68a0a867d830a7f442d2a6R538

Tests added are at:
https://github.com/django/django/pull/8528/files#diff-09569af128ce9190c46eaae0d7330a23R1381

MatthewWilkes commented on django/db/models/sql/compiler.py:
```diff
@@ -622,21 +627,22 @@ def get_distinct(self):
         for name in self.query.distinct_fields:
             parts = name.split(LOOKUP_SEP)
-            _, targets, alias, joins, path, _ = self._setup_joins(parts, opts, None)
+            field, targets, alias, joins, path, _, transform_function = self._setup_joins(parts, opts, None)
```
Good catch, it's rebased a few times over the versions, I have no idea if this was a careless copy/paste or a bad rebase, but it can go.

MatthewWilkes commented on django/db/models/sql/query.py:
```diff
@@ -1471,7 +1511,7 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True,
             if filtered_relation:
                 filtered_relation.path = joins[:]
-        return JoinInfo(final_field, targets, opts, joins, path)
+        return JoinInfo(final_field, targets, opts, joins, path, final_transformer)
```
I've added a description of the transform function to the docstring but I didn't include information about the named tuple, perhaps I should have done this as part of #8840 but I think maybe if we think this should reference JoinInfo that can be a small improvement rather than adding to this monster of a ticket?

carltongibson replied:
I'd rather see the docstring updated in this one: just so it's not forgotten. With the change, the paragraph beginning `Return...` isn't right. 

My worry is if we leave it then it'll be forgotten.

timgraham replied:
Please don't suggest doing unrelated cleanups along with bug fix or feature work. We try to avoid that as it confuses things when studying a patch.

MatthewWilkes replied:
I added the transform function info to the 'Return...' paragraph so that it won't be incorrect, I agree that it's not the clearest docstring in the world an that improving it would be worthwhile, but we already split out the JoinInfo cleanup from this ticket and I don't really want to reintroduce it.

carltongibson replied:
Thanks @MatthewWilkes. It was only that sentence/paragraph that needs fixing here, since it's adding the new value that will be unpacked. Any other clean up can wait.

MatthewWilkes commented on docs/releases/2.1.txt:
```diff
@@ -216,6 +219,9 @@ Database backend API
   :exc:`django.db.NotSupportedError`.
 
+* The ``DatabaseOperations.distinct_sql()`` method now returns a tuple
+  of SQL and parameters, instead of a SQL string.
```
It's worth adding, you're right.

MatthewWilkes commented on docs/ref/models/querysets.txt:
```diff
@@ -535,6 +535,18 @@ The ``values()`` method also takes optional keyword arguments,
+
+The same is true of the built-in lookups, such as ``__date`` or ``__month`` on
+:class:`~django.db.models.DateTimeField`, as described in :doc:`database-functions`.
+
```
I see what you mean. I'll add a note in the subtleties list but I think I'd be -1 on making a big deal of it in the main narrative as it seems like such an edge case.

MatthewWilkes commented on tests/backends/base/test_operations.py:
```diff
@@ -17,7 +17,7 @@ def setUp(self):
         msg = 'DISTINCT ON fields is not supported by this database backend'
         with self.assertRaisesMessage(NotSupportedError, msg):
-            self.ops.distinct_sql(['a', 'b'])
+            self.ops.distinct_sql('a', 'b')
```
Strongly agree.

MatthewWilkes commented:
Thanks for your comments, @carltongibson! I've pushed a revision that should address the individual line comments. I'm also unsure about the approach for the readme and I haven't added examples in there yet but am open to doing so.

I think that maybe if  we're going to include how to order by a nested JSON key in the documentation it would need to include how to set up the appropriate indexes?

timgraham commented:
Generally any examples should be included in the documentation and linked to from the release notes rather than included in the release notes.

timgraham commented:
buildbot, test on oracle.

MatthewWilkes commented:
Should I rebase this to fix the conflict in the release notes, or will the committer do that change? I'm just aware that it'll trigger a dozen builds and remove the confirmation that it passed on oracle.

carltongibson commented:
@MatthewWilkes I think this was good to go. We should be able to rebase. Thank you!

timgraham commented on tests/expressions/tests.py:
```diff
@@ -1357,6 +1357,40 @@ def test_raise_empty_expressionlist(self):
             ExpressionList()
 
 
+class FLookupTests(TestCase):
```
Is the class name correct? I don't see any F expressions -- maybe I misunderstood what "FLookup" means.

MatthewWilkes replied:
This was from jarshwah's initial hint in the ticket. I agree, your suggested name is better.

timgraham commented on docs/ref/models/querysets.txt:
```diff
@@ -580,6 +593,23 @@ A few subtleties that are worth mentioning:
+    >>> from django.db.models import CharField, Count
+    >>> from django.db.models.functions import Lower
+    >>> CharField.register_lookup(Lower, 'lower')
+    >>> Blog.objects.values('author__name__lower').annotate(entries=Count('entry'))
```
I think these examples need correcting as `Blog` doesn't have an `author` field.

MatthewWilkes replied:
That's a good point. I'd expanded on an example higher in the file, which was fixed in the mainline as #9601. I'll apply the same change here.

timgraham commented on docs/ref/models/expressions.txt:
```diff
@@ -64,10 +64,11 @@ Some examples
     Company.objects.order_by(Length('name').desc())
-
+    # or using the double underscore syntax.
+    Company.objects.order_by('products__count')
```
Does `__count` need to be registered as a transform for this to work? I guess it's a bit unclear what `products` refers to since the `Company` model isn't defined as far as I see.

MatthewWilkes replied:
A good point. I'm going to make this use `name__length` instead. It's a bit clearer, as name is a normal field rather than a relationship and we have a similar example in the 1.10 release notes.

MatthewWilkes commented:
I've pushed a change that I believe addresses the comments from 2 days ago and 20 minutes ago. To be clear, I agree with the changes you've made directly: thanks!
